### ISSUE TITLE
Can't build Windows version locally with dune

### ISSUE THREAD

**Issue by jfehrle** [coq/coq#16095]
After a full local build with dune for Windows, I find the prelude is unusable.  I wanted to debug a Windows-specific issue.  What should I do?

```
$ dune exec -- dev/shim/coqide-prelude
Error:
create_process(C:\proj\coq2\_build/default/dev/shim/coqide-prelude): Exec format error

$ file ./_build/default/ide/coqide/coqide_main.exe       (* IIUC this is what the prelude uses *)
./_build/default/ide/coqide/coqide_main.exe: PE32+ executable (console) x86-64, for MS Windows

$ file `which coqide`     (* version set up by Coq platform, same format, this runs just fine *)
/home/jfehr/.opam/__coq-platform.2022.04.0~8.15~2022.04/bin/coqide: PE32+ executable (console) x86-64, for MS Windows
```

`coqide_main.exe` is indeed runnable, though it dies showing this dialog:

![image](https://user-images.githubusercontent.com/1253341/170894002-b1d9668b-ab32-4500-be2c-79852304f24b.png)

Any idea why I get the fatal error below?

```
$ ./_build/default/ide/coqide/coqide_main.exe
Gtk-Message: GtkDialog mapped without a transient parent. This is discouraged.
Fatal error: exception Unix.Unix_error(Unix.ENOENT, "create_process", "coqidetop.opt")
```

BTW, why is the program named `coqide_main.exe` instead of, say, `coqide.exe`?

**Comment by Alizter** [comment]
The name of the program is correct. That is the exe that gets produced in the _build/default directory. If you want to run what `dune exec` actually runs, you need to look in `_build/install/default/bin`.

**Comment by Alizter** [comment]
I would suggest doing `dune build theories/Init @check coqide.install` to get most of it build. I am not sure why the coqide shim didn't work here.

**Comment by jfehrle** [comment]
Worse:

```
$ dune build theories/Init @check coqide.install
File "tools/flock/dune", line 5, characters 9-14:
5 |   (names flock))
             ^^^^^
flock.c: In function ‘coq_flock’:
flock.c:24:19: error: ‘LOCK_SH’ undeclared (first use in this function)
   24 |     operation_f = LOCK_SH;
      |                   ^~~~~~~
flock.c:24:19: note: each undeclared identifier is reported only once for each function it appears in
flock.c:27:19: error: ‘LOCK_EX’ undeclared (first use in this function)
   27 |     operation_f = LOCK_EX;
      |                   ^~~~~~~
flock.c:30:19: error: ‘LOCK_UN’ undeclared (first use in this function)
   30 |     operation_f = LOCK_UN;
      |                   ^~~~~~~
flock.c:37:13: warning: implicit declaration of function ‘flock’ [-Wimplicit-function-declaration]
   37 |   int res = flock(fd_i, operation_f);
      |             ^~~~~
```

**Comment by Alizter** [comment]
OK so lots of things going wrong here at the same time:
1. dune exec doesn't do bash.
Have a look at the shim rule:
```
(rule
 (targets coqide-prelude)
 (deps
  ; without this if the gtk libs are not available dune can try to use
  ; coqide from PATH instead of giving a nice error
  ; there is no problem with the other shims since they don't depend on optional build products
  %{project_root}/ide/coqide/coqide_main.exe

  %{bin:coqqueryworker.opt}
  %{bin:coqtacticworker.opt}
  %{bin:coqproofworker.opt}
  %{project_root}/theories/Init/Prelude.vo
  %{project_root}/coqide-server.install
  %{project_root}/coqide.install)
 (action
  (with-stdout-to coqide-prelude
   (progn
    (echo "#!/usr/bin/env bash\n")
    (bash "echo '\"$(dirname \"$0\")\"/%{bin:coqide} -I \"$(dirname \"$0\")/%{project_root}/../install/default/lib\" -coqlib \"$(dirname \"$0\")\"/%{project_root} \"$@\"'")
    (run chmod +x %{targets})))))
```
As you can see there is "bash" component of the action component. Dune cannot run this on Windows.

2. flock and some other stuff is also subject to Windows unavailability.

This one can be solved easily however. We will drop flock very soon, but the entire thing could have been avoided since I tried to give a "catch all" command.

What I should have said is:
```
dune build theories/Init coqide.install
```
The `@check` target was to compile all the ml in the coq repo, but that was a little too much.

---

In short, the coqide-prelude shim will not work for Windows. Instead, you will have to `dune build theories/Init` and `dune build coqide.install`. Then you may `dune exec -- coqide.exe`.

**Comment by ejgallego** [comment]
> As you can see there is "bash" component of the action component. Dune cannot run this on Windows.

That's not exactly the problem. Dune runs the rule, what it can't run is the generated shell script file that starts with the shebang.

**Comment by jfehrle** [comment]
@Alizter Just `coqide`, not `coqide.exe`.   The GUI window comes up, but I get a "coqidetop died" dialog.  Looks like the backslash in the coqidetop path should be a forward slash.  If I run the Coq platform-installed version, the filename is entirely in Windows format: 
`C:\cygwin64_coq_platform\home\jfehr\.opam\__coq-platform.2022.04.0~8.15~2022.04\bin\coqidetop.opt.exe`.

```
$ dune exec -- coqide.exe
Error: Program "coqide.exe" not found!

$ dune exec -- coqide -debug
(Spawn  ,188) EXEC: _build/install/default/bin\coqidetop.opt.exe -main-channel 127.0.0.1:55777:55778 -control-channel 127.0.0.1:55775:55776 --xml_format=Ppcmds -async-proofs off
[DEBUG] Start eval_call Init None
[DEBUG] End eval_call
[DEBUG] End of Coqide.main
[DEBUG] Signal ignored (normal if Win32)
[DEBUG] Coqtop reader failed, resetting: Coq.TubeError("ERR")

$ _build/install/default/bin\coqidetop.opt.exe
bash: _build/install/default/bincoqidetop.opt.exe: No such file or directory
$ _build/install/default/bin/coqidetop.opt
Fatal error: ideslave communication channels not set.
```

When I hard-code `C:\proj\coq2\_build\install\default\bin\coqidetop.opt` in the EXEC, that still fails, and same if I append `.exe`.  (The filename has `.exe`)  Not sure what to try next.

FWIW, I verified that the dune version has os_type = "Win32".
Also FWIW, my shell is cygwin, not mingw (if the latter is even possible).

**Comment by jfehrle** [comment]
If I pass `-d all` to coqtopide, I see messages from it, but it appears to die abruptly with no stacktrace.  It does get as far as printing the `SENT` message.  Adding a try block around the main routine in `Idetop` as shown I get no result, so it must die abruptly somehow.  Could it be missing files somehow, e.g. dynamic code that's not found?

```
try
  let open Coqtop in
  let custom = {
      parse_extra = islave_parse ;
      usage = coqidetop_specific_usage;
      init_extra = islave_init;
      run = loop;
      initial_args = islave_default_opts } in
  start_coq custom
with e -> Printf.eprintf "%s%!" (Printexc.to_string e); Printexc.print_backtrace stderr
```

```
$ dune exec -- coqide
pid = 1260
Entered idetop
[pid 232] <call val="Init">
  <option val="none"/>
</call>
[pid 232] <-- Init None
[pid 232] --> GOOD 1
SENT --> <value val="good">
  <state_id val="1"/>
</value>
```

**Comment by jfehrle** [comment]
Ah, the server process doesn't die, it's just waiting to read the next XML message.  I can see the process in the Windows Task Manager.  But no idea why CoqIDE can't receive the message or why the behavior should be any different when I do a local build with dune.

**Comment by ejgallego** [comment]
Duplicate of #14547
